### ISSUE TITLE
Geometry cutter on the GPU

### ISSUE THREAD

**Issue by agirault** [Kitware/vtk-js#1967]
### Need

Display the contour of a 3d model when intersected with a plane, to display on top of an image data slice for example.

### Feature

- **suggestion**: create a mapper that can take a `vtkPolyData` and a plane equation (or arbitrary equation) to compute its intersection on the GPU
- **alternative**: can currently do it on the CPU with `vtkCutter` for each slice, but that's slow

**Comment by martinken** [comment]
I would do this only for WebGPU (WebGL will be a problem with Apple only supporting WebGL 1) And I would do it by creating a compute shader that takes triangles and cuts them into lines based on the plane equation.

Right now WebGPU is probably broken (just guessing as they are continuously changing WebGPU leading up to release) and probably needs some attention to get it up to the current version in chrome canary. That would be my suggestion for a first step as it will fix also help you get familiar with WebGPU.

**Comment by martinken** [comment]
Also note that WebGPU is slated to be released in chrome mainline I think in another 4 releases. It may not happen bet we need to freshen it up soon anyhow. 

<blockquote class="twitter-tweet"><p lang="und" dir="ltr"><a href="https://t.co/dMtn7dBegN">https://t.co/dMtn7dBegN</a></p>&mdash; WebGPU (@WebGPU) <a href="https://twitter.com/WebGPU/status/1431184221448687620?ref_src=twsrc%5Etfw">August 27, 2021</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>

**Comment by agirault** [comment]
Thanks @martinken. Any opinion whether we should extend an existing mapper with that property (similarly to `VolumeMapper.addClippingPlane` ?

**Comment by martinken** [comment]
Yes, it would be a property on the mapper just like clipping planes but only one of them.

**Comment by ebrahimebrahim** [comment]
I am going to be tackling this issue!

**Comment by ebrahimebrahim** [comment]
Since WebGPU is still in an experimental phase, I think I am going to keep my focus on WebGL2 here. WebGL2 is really pervasive right now, so even if it's being replaced I don't think it's going anywhere soon. What do you think @martinken ?

Ideally, eventually, we want every feature to be supported in both backends, right? Is that the goal? @floryst

**Comment by martinken** [comment]
I am only implementing new features on WebGPU, I doubt we have the resources to do both for any extended time. But once you've done it in WebGL2 maybe it will be a bit quicker to do it in WebGPU.

**Comment by ebrahimebrahim** [comment]
Thanks for the advice @martinken , I think I might at least start on the WebGL2 side because I have some background with OpenGL and I have no background with Vulkan (which I am guessing WebGPU is more similar to). Then I'll have learned more about the particulars of vtk.js rendering, and I might stand a better chance learning about WebGPU for future tasks.

**Comment by sedghi** [comment]
I was wondering if there is any worthy update on the status of a GPU based geometry cutter on WebGL2 or WebGPU since last year. Thanks!

**Comment by ebrahimebrahim** [comment]
Ah, this task should be unassigned from me.

@sedghi No update. I moved out of active vtk.js development and never got back to update this.

Feel free to take this on if you want :slightly_smiling_face:

**Comment by daker** [comment]
@finetjul I've got a working GPU based vtkCutterMapper(plane only), do you think it's worth pushing?

**Comment by finetjul** [comment]
@daker if you could do it as something that can be "plugged" (composition) onto the existing vtkPolyDataMapper, that would be great.
